### PR TITLE
Fix issue where resizing an image doesn't always give an exact width/height

### LINKED ISSUE #667: Resizing an image doesn't always give an exact width/height
I have been using `image` to resize images for the purpose of creating thumbnails. I noticed that if I resize an image with an automatic dimension (e.g. pass `std::u32::MAX` as height so the width is always exactly what I request), the dimension I do pass in is sometimes one lower than what I would expect.

For example, if I run this code on this [image](https://i.imgur.com/mSl1LZY.jpg):

```rust
fn resize_image(image: &DynamicImage) {
    let height = 880;
    let test_image = image.resize(::std::u32::MAX, height, FilterType::Lanczos3);
    println!("{}", test_image.height());
}
```

I would expect the resulting height of `test_image` to be 880, but instead it is 879.

### PR DESCRIPTION
**Pull request by shssoichiro**
Also refactors out common resizing logic from `resize` and `resize_to_fill`

It seems that this was due to insufficient precision in the `f32` type which was causing rounding errors in some cases. `f64` seems to have enough precision to handle this accurately, at least in the cases I tested.

Closes #667

### CODE REVIEW

File: src/dynimage.rs [hunk 1]
```diff
@@ -324,6 +302,25 @@ impl DynamicImage {
         }
     }
 
+    fn resize_dimensions(&self, nwidth: u32, nheight: u32, fill: bool) -> (u32, u32) {
```
Comment (calum) [inline src/dynimage.rs]: I think it would be beneficial to have a unit test for this function. However, I'm not sure where it should go. There aren't many tests written at the bottom of this file.

Comment (bvssvni) [reply src/dynimage.rs]: The bottom at the file is fine, I think. Could put it here: https://github.com/PistonDevelopers/image/blob/master/src/dynimage.rs#L764

**Comment by shssoichiro** [comment]
Any update on this? I've added unit tests around the resizing logic.

**Comment by bvssvni** [comment]
Looks good to me.

**Comment by HeroicKatora** [comment]
Wouldn't it be better to use integer logic with `u64` instead? As in 
`iwidth / iheight < nwidht / nheight` <=> `iwhdth * nheight < nwidth*iheight`

`f64` with 53 bits of precision could be enough though. Compared to the old, the input arguments are surely represented without error and the division result is only be off by at most one `ulp`. Maybe this should be checked more thoroughly regardless.

**Comment by shssoichiro** [comment]
I do like that better, if nothing else because integer math is generally faster. We still have to use float division to calculate the ratios, but can safely use u64's for calculating the new dimensions.

**Comment by shssoichiro** [comment]
I also realized I could simplify two of the expressions e.g. `(width as u64 * nwidth as u64 / width as u64) as u32` to just `nwidth`. So I pushed another new commit to include that.

**Comment by HeroicKatora** [comment]
Should we handle the case that calling with `fill = True` fails the cast to `u32`, i.e. the size got too big?

**Comment by shssoichiro** [comment]
I agree. I've added tests around fill with and without overflow, and updated the code to handle the overflow case by bounding it to `u32::MAX`. I'm not sure that this code is the cleanest, but it does handle the edge cases.

**Comment by shssoichiro** [comment]
I had to push a commit to fix a compile error if pnm feature is disabled; I was trying to pull this branch into a project where only png and jpeg are enabled, and encountered that. I realize it's not related to this issue, but I needed this branch to work.

File: src/dynimage.rs [hunk 2]
```diff
@@ -747,6 +725,31 @@ pub fn guess_format(buffer: &[u8]) -> ImageResult<ImageFormat> {
     )
 }
 
+fn resize_dimensions(width: u32, height: u32, nwidth: u32, nheight: u32, fill: bool) -> (u32, u32) {
```
Comment (bvssvni) [inline src/dynimage.rs]: Can you add some comments? Explain e.g. `fill`.

**Comment by bvssvni** [comment]
Merging.
